Avoid futures unordered to receive replica updates

antiguru opened the pull request:
Switch storage replicas to use a shared channel for sending responses back to the controller. This avoids the need for `FuturesUnordered` to select across a dynamic set of channels.

Part of #28720.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

teskje commented on src/storage-controller/src/instance.rs:
```diff
@@ -65,6 +62,8 @@ pub(crate) struct Instance<T> {
     now: NowFn,
     /// Responses ready for delivery on the next call to [`Instance::recv`].
     ready_responses: VecDeque<StorageResponse<T>>,
```
Is there a reason to keep this? Seems like we can just push responses to `instance_response_tx` instead.

antiguru replied:
Removed

teskje commented on src/storage-controller/src/instance.rs:
```diff
@@ -65,6 +62,8 @@ pub(crate) struct Instance<T> {
     /// Responses ready for delivery on the next call to [`Instance::recv`].
     ready_responses: VecDeque<StorageResponse<T>>,
+    /// A receiver for responses from replicas.
+    instance_response_tx: mpsc::UnboundedSender<StorageResponse<T>>,
```
Nit: I'd name this just `response_tx`, since we're already in the context of an instance here.

teskje commented on src/storage-controller/src/instance.rs:
```diff
@@ -299,46 +276,29 @@ where
-        if self.command_tx.send(command).is_err() {
-            self.failed = true;
-        }
+        /// Send failures ignored, we'll check for failed replicas separately.
```
```suggestion
        // Send failures ignored, we'll check for failed replicas separately.
```

teskje reviewed (commented):
I like the idea. I commented with some simplifications I think we can do.

We should do a similar change for the compute controller as well, to keep the two as similar as possible. But I'd prefer to defer that until after the compute controller decoupling is merged, to avoid conflicts.

antiguru commented:
Thanks for the review! I addressed the feedback. Yes, for the compute controller we can apply a  similar pattern. Happy to wait with it until @teskje's changes land.

teskje reviewed (approved):
LGTM!

bosconi commented:
@teskje if there isn't already an issue for the corresponding change in the compute controller, is it worth having one to track it, and if so would you mind creating it?

teskje commented:
Yep! Created https://github.com/MaterializeInc/database-issues/issues/8587 and assigned @antiguru.
